vscode-lean4/src/leanclient.ts

```diff
@@ -98,30 +104,71 @@ export class LeanClient implements Disposable {
             await this.stop()
         }
 
-        this.executable = this.storageManager.getLeanPath();
-        if (!this.executable) this.executable = executablePath();
+        this.toolchainPath = this.storageManager.getLeanPath();
+        if (!this.toolchainPath) this.toolchainPath = toolchainPath();
         const version = this.storageManager.getLeanVersion();
         const env = addServerEnvPaths(process.env);
         if (serverLoggingEnabled()) {
             env.LEAN_SERVER_LOG_DIR = serverLoggingPath()
         }
 
-        let options = ['--server']
-        if (version) {
-            // user is requesting an explicit version for this workspace.
-            options = ['+' + version, '--server']
+        let executable = (this.toolchainPath) ? join(this.toolchainPath, 'bin', 'lake') : 'lake';
+
+        // check if the lake process will start.
+        let useLake = lakeEnabled();
+        if (useLake && this.folderUri) {
+            const lakefile = Uri.joinPath(this.folderUri, 'lakefile.lean').toString()
+            if (!fs.existsSync(new URL(lakefile))) {
+                useLake = false;
+            }
+        }
+
+        // This is a faster way of finding out lake doesn't work in the current workspace.
+        // The LanguageClient is much slower because it does 10 retries and everything.
+        if (useLake) {
+            // First check we have a version of lake that supports "lake serve"
+            const versionOptions = version ? ['+' + version, '--version'] : ['--version']
+            const lakeVersion = await batchExecute(executable, versionOptions, this.folderUri?.fsPath, null);
+            const actual = this.extractVersion(lakeVersion)
+            if (actual.compare('3.0.0') >= 0) {
+                const expectedError = 'Watchdog error: Cannot read LSP request: Stream was closed\n';
+                const serveOptions =  version ? ['+' + version, 'serve'] : ['serve'];
+                const rc = await testExecute(executable, serveOptions, this.folderUri?.fsPath, this.outputChannel, true, expectedError);
+                if (rc !== 0) {
+                    const failover = 'Lake failed, using lean instead.'
+                    console.log(failover);
+                    if (this.outputChannel) this.outputChannel.appendLine(failover);
+                    useLake = false;
+                }
```

Comment:
I think we can drop this check now.  `lake serve` falls back to `lean --server` now if the `lakefile.lean` doesn't compile.

Reply:
Do you know what version of lake does that?  I tried this version: leanprover/lean4:nightly-2022-02-10, with a syntax error in the lakefile.lean and I get these retry errors and no lean server running:
```
client starting
vscode-lean4/src/leanclient.ts:257
client has stopped or it failed to start
vscode-lean4/src/leanclient.ts:262
client starting
vscode-lean4/src/leanclient.ts:257
client has stopped or it failed to start
vscode-lean4/src/leanclient.ts:262
client starting
vscode-lean4/src/leanclient.ts:257
client has stopped or it failed to start
vscode-lean4/src/leanclient.ts:262
client starting
vscode-lean4/src/leanclient.ts:257
client has stopped or it failed to start
vscode-lean4/src/leanclient.ts:262
client starting
vscode-lean4/src/leanclient.ts:257
client has stopped or it failed to start
vscode-lean4/src/leanclient.ts:262
```

Reply:
Ohh, apparently the bundled lake hasn't been updated since https://github.com/leanprover/lake/pull/52 was merged.  Which means we should probably keep this check for now (and gate it on the lake version in the future).

Reply:
Since this was the only commit not merged upstream, it felt safe to do so now. /cc @tydeu